phoenix-core/src/main/java/org/apache/phoenix/hbase/index/builder/IndexBuildManager.java

```diff
@@ -88,27 +111,102 @@ public void getIndexUpdates(ListMultimap<HTableInterfaceReference, Pair<Mutation
       IndexMetaData indexMetaData) throws Throwable {
     // notify the delegate that we have started processing a batch
     this.delegate.batchStarted(miniBatchOp, indexMetaData);
-
+    CachedLocalTable cachedLocalTable =
+            preScanAllRequiredRows(
+                    mutations,
+                    (PhoenixIndexMetaData)indexMetaData,
+                    this.regionCoprocessorEnvironment);
     // Avoid the Object overhead of the executor when it's not actually parallelizing anything.
     for (Mutation m : mutations) {
-      Collection<Pair<Mutation, byte[]>> updates = delegate.getIndexUpdate(m, indexMetaData);
+      Collection<Pair<Mutation, byte[]>> updates = delegate.getIndexUpdate(m, indexMetaData, cachedLocalTable);
       for (Pair<Mutation, byte[]> update : updates) {
         indexUpdates.put(new HTableInterfaceReference(new ImmutableBytesPtr(update.getSecond())), new Pair<>(update.getFirst(), m.getRow()));
       }
     }
   }
 
+  @VisibleForTesting
+  public static CachedLocalTable preScanAllRequiredRows(
+          Collection<? extends Mutation> dataTableMutationsWithSameRowKeyAndTimestamp,
+          final PhoenixIndexMetaData indexMetaData,
+          RegionCoprocessorEnvironment regionCoprocessorEnvironment) throws IOException {
```

Comment:
@gjacoby126 , ok ,thank you for the review,  I use RegionCoprocessorEnvironment  here because the convenience for modifying tests, I will modify according your suggestions.